server/session/handler_anvil.go

```diff
@@ -0,0 +1,292 @@
+	// the durability of the result by the durability delta.
+	var cost, count int
+	for ; delta > 0 && count < material.Count(); count, delta = count+1, min(result.MaxDurability()-result.Durability(), result.MaxDurability()/4) {
+		result = result.WithDurability(result.Durability() + delta)
+		cost++
+	}
+	return result, cost, count, nil
+}
+
+// repairWithDurable is a helper function that repairs an item with another durable item stack.
+func repairWithDurable(input item.Stack, durable item.Stack, result item.Stack) (item.Stack, int) {
+	durability := input.MaxDurability() - (input.Durability() + (durable.Durability() + input.MaxDurability()*12/100))
+	if durability < 0 {
+		durability = 0
+	}
+
+	// Ensure the durability increase is less than the input's current damage.
+	var cost int
+	if durability < input.MaxDurability()-input.Durability() {
+		result = result.WithDurability(durability)
+		cost += 2
+	}
+	return result, cost
+}
+
+// mergeEnchantments merges the enchantments of the material item stack onto the result item stack and returns the result
+// item stack, booleans indicating whether the enchantments had any compatible or incompatible enchantments, and the cost.
+func mergeEnchantments(input item.Stack, material item.Stack, result item.Stack, cost int, enchantedBook bool) (item.Stack, bool, bool, int) {
+	var hasCompatible, hasIncompatible bool
+	for _, enchant := range material.Enchantments() {
+		// First ensure that the enchantment type is compatible with the input item.
+		enchantType := enchant.Type()
+		compatible := enchantType.CompatibleWithItem(input.Item())
+		if _, ok := input.Item().(item.EnchantedBook); ok {
+			compatible = true
+		}
+
+		// Then ensure that each input enchantment is compatible with this material enchantment.
+		for _, otherEnchant := range input.Enchantments() {
+			otherType := otherEnchant.Type()
+			if enchantType != otherType && !enchantType.CompatibleWithEnchantment(otherType) {
+				compatible = false
+				cost++
+			}
+		}
+
+		// Skip the enchantment if it isn't compatible with enchantments on the input item.
+		if !compatible {
+			hasIncompatible = true
+			continue
+		}
+		hasCompatible = true
+
+		// First check if we have an enchantment of the same type on the input item. If so, record it's level.
+		var firstLevel int
+		if firstEnchant, ok := input.Enchantment(enchantType); ok {
+			firstLevel = firstEnchant.Level()
+		}
+
+		// Then calculate the result level by getting the maximum of the first level and this enchantment's
+		// level. If the first level is equal to the enchantment's level, increase the result level by one.
+		resultLevel := max(firstLevel, enchant.Level())
+		if firstLevel == enchant.Level() {
+			resultLevel = firstLevel + 1
+		}
+
+		// If the result level is greater than the enchantment's maximum level, set it to the maximum level
+		// instead.
+		if resultLevel > enchantType.MaxLevel() {
+			resultLevel = enchantType.MaxLevel()
```

Comment:
Should probably be considered incompatible; vanilla will not allow combining if all enchantments on the material item are already maxed on the input item.

Reply:
Similarly if all enchantments on the material item are lower in level than on the input item, combining will be blocked.